include/sys/queued_operation.h

```diff
@@ -0,0 +1,278 @@
+ * constant-time insertion and is LIFO.
+ */
+#define QUEUED_OPERATION_PRIORITY_PREPEND ((int)QUEUED_OPERATION_PRIORITY_MASK + 2)
+
+/**
+ * @brief Identify the region of async_notify flags available for
+ * containing services.
+ *
+ * Bits of the flags field of the async_notify structure contained
+ * within the queued_operation structure at and above this position
+ * may be used by extensions to the async_notify structure.
+ *
+ * These bits are intended for use by containing service
+ * implementations to record client-specific information.  The bits
+ * are cleared by async_notify_validate().  Use of these does not
+ * imply that the flags field becomes public API.
+ */
+#define QUEUED_OPERATION_EXTENSION_POS (QUEUED_OPERATION_PRIORITY_POS + QUEUED_OPERATION_PRIORITY_BITS)
+
+/**
+ * @brief Base object providing state for an operation.
+ *
+ * Instances of this should be members of a service-specific structure
+ * that provides the operation parameters.
+ */
+struct queued_operation {
+	/** @internal
+	 *
+	 * Links the operation into the operation queue.
+	 */
+	sys_snode_t node;
+
+	/**
+	 * @brief Notification configuration.
+	 *
+	 * This must be initialized using async_notify_init_callback()
+	 * or its sibling functions before an operation can be passed
+	 * to queued_operation_submit().
+	 */
+	struct async_notify notify;
+};
+
+/**
+ * @ brief Table of functions used by a queued operation manager.
+ */
+struct queued_operation_functions {
+	/**
+	 * @brief Function used to verify an operation is well-defined.
+	 *
+	 * When provided this function is invoked by
+	 * queued_operation_submit() to verify that the operation
+	 * definition meets the expectations of the service.  The
+	 * operation is acceptable only if a non-negative value is
+	 * returned.
+	 *
+	 * If not provided queued_operation_submit() will assume
+	 * service-specific expectations are trivially satisfied, and
+	 * will reject the operation only if async_notify_validate()
+	 * fails.  Because that validation is limited services should
+	 * at a minimum verify that the extension bits have the
+	 * expected value (zero, when none are being used).
+	 *
+	 * @param mgr the service that supports queued operations.
+	 *
+	 * @param op the operation being considered for suitability.
+	 *
+	 * @return the value to be returned from queued_operation_submit().
+	 */
+	int (*validate)(struct queued_operation_manager *mgr,
+			struct queued_operation *op);
+
+	/**
+	 * @brief Function to transform a generic notification
+	 * callback to its service-specific form.
+	 *
+	 * The implementation should cast cb to the proper signature
+	 * for the service, and invoke the cast pointer with the
+	 * appropriate arguments.
+	 *
+	 * @param mgr the service that supports queued operations.
+	 *
+	 * @param op the operation that has been completed.
+	 *
+	 * @param cb the generic callback to invoke.
+	 */
+	void (*callback)(struct queued_operation_manager *mgr,
```

Comment:
shouldn't it contain result? I know that it is written inside async_notify in operation but it will definitely be needed and fetching it inside callback seems strange as it is already available when this callback is invoked.

Reply:
I don't think it's necessary.  The intermediary translation function provided by the service can extract the result if passing it to the end user as a distinct argument is useful.  In a probably common case where the completed operation is simply appended to another list for processing in another context it isn't needed.